---
title: "R Assignment"
author: "Akhil"
date: "10/18/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

# Data Inspection

#Read a txt file, named "fang_et_al_genotypes.txt"

fang<- read.delim("fang_et_al_genotypes.txt")


For Rows

length(readLines("fang_et_al_genotypes.txt"))-2783


For columns

ncol(fang)-986

install.packages("MODIS")

#File size

file.size("fang_et_al_genotypes.txt", units = "B")-11051939bytes

#The file permissions, printed in octal

file.mode("fang_et_al_genotypes.txt")-666

#last modified time stamp

file.mtime("fang_et_al_genotypes.txt")-2019-10-17 19:10:01 CDT

#For snp_position.txt
#Read a txt file, named "snp_position.txt"

 snp<- read.delim("snp_position.txt")
 
#For Rows

length(readLines("snp_position.txt"))-984

#For columns
ncol(snp)-15

file.size("snp_position.txt", units = "B")-82763bytes

#The file permissions, printed in octal

file.mode("snp_position.txt")- "666"

#last modified time stamp

file.mtime("snp_position.txt")- "2019-10-17 19:16:30 CDT"

## Data processing

#Creation of data frame snp_position.txt without column 2 but 1,3,4

Snp1 = select(snp, 1,-2,3,4)

#Creation of data frame fang_genotype_et_al.txt with group "ZMMLR", "ZMMMR", "ZMMIL"

Maize_ <- filter(fang, Group %in% c("ZMMLR", "ZMMMR", "ZMMIL"))

#Creation of data frame fang_genotype_et_al.txt with group "ZMPBA", "ZMPIL", "ZMPJA"

Teosinte_ <- filter(fang, Group %in% c("ZMPBA", "ZMPIL", "ZMPJA"))

#Creation of data frame  without column 2 & 3 but 1,4:986

col_Maize=select(Maize_, 1,-2,-3,4:986)

col_Teosinte=select(Teosinte_1,-2,-3,4:986)


#Without Headers

colnames(col_Maize) <- c()

colnames(col_Teosinte) <- c()

#Transposed data

transposed_maize <- t(col_Maize)

transposed_Teosinte <- t(col_Teosinte)

#Merged data frame for Maize

Join_Maize<-merge(Snp1, transposed_maize, by=NULL, all.x=TRUE)

#Merged data frame for Teosinte

Join_Teosinte<-merge(Snp1, transposed_Teosinte, by=NULL, all.x=TRUE)

#the large  data is slowing down R. Hence taking into account only the needed first 15columns(SNP, chromosome, position) of the joined file

Join_Maize = select(Join_Maize, 1:15)

Join_Teosinte = select(Join_Teosinte, 1:15)


#replace "?/?" to "?"

Join_Maize <- data.frame(lapply(Join_Maize, as.character), stringsAsFactors=FALSE)

Join_Maize <- data.frame(sapply(Join_Maize,function(x) {x <- gsub("?/?","?",x,fixed=TRUE)}))

Join_Teosinte <- data.frame(lapply(Join_Teosinte, as.character), stringsAsFactors=FALSE)

Join_Teosinte<- data.frame(sapply(Join_Teosinte,function(x) {x <- gsub("?/?","?",x,fixed=TRUE)}))

for (i in 1:10) {

  maize <- filter(maize_fang_et_al_genotypes, Chromosome == i)
  
  maize <- arrange(maize, Position)
  
  outpath <- "C:/Users/akhil/Desktop/R/R Assignment/Files/"
  
  nam <- sapply(
  
    names(maize),function(x){
    
      paste("maize", i, ".csv", sep='')
    })
  out_filePath <- sapply(nam, function(x){
  
    paste(outpath, x, sep='/')})
    
  write.csv(maize, file=out_filePath[i]) # for maize increasing data

#above code is not working. Hence moving on  
  
  
#replace "?" to "-"

Join_Maize2 <- data.frame(lapply(Join_Maize, as.character), stringsAsFactors=FALSE)

Join_Maize2 <- data.frame(sapply(Join_Maize2,function(x) {x <- gsub("?","-",x,fixed=TRUE)}))

Join_Teosinte2 <- data.frame(lapply(Join_Teosinte, as.character), stringsAsFactors=FALSE)

Join_Teosinte2<- data.frame(sapply(Join_Teosinte2,function(x) {x <- gsub("?","-",x,fixed=TRUE)}))


#For sorting in decreasing order of position

Join_Teosinte_decposition <- arrange(Join_Teosinte2, desc(Position))

#For selecting first 10 files of position sorted

Join_Teosinte_decposition <- Join_Teosinte_decposition[1:10,]

#for saving file

write.csv(Join_Teosinte_decposition,'Join_Teosinte_decposition')



#For sorting in decreasing order of position

Join_Maize2_decposition <- arrange(Join_Maize2, desc(Position))

#For selecting first 10 files of position sorted

Join_Maize2_decposition <- Join_Maize2_decposition[1:10,]

write.csv(Join_Maize2_decposition,'Join_Maize2_decposition')



#For sorting in increasing order of position

Join_Maize_incposition <- arrange(Join_Maize, Position)

#For selecting first 10 files of position sorted

Join_Maize_incposition <- Join_Maize_incposition[1:10,]

#for saving file

write.csv(Join_Maize_incposition,'Join_Maize_incposition')



#For sorting in increasing order of position

Join_Teosinte_incposition <- arrange(Join_Teosinte, Position)

#For selecting first 10 files of position sorted

Join_Teosinte_incposition <- Join_Teosinte_incposition[1:10,]

#for saving file

write.csv(Join_Teosinte_incposition,'Join_Teosinte_incposition')




#COde not working to push and save as 10 seprate files. hence moving on.

# Graphs

install.packages("tidyverse") | library(tidyverse)

 install.packages("reshape2") |library(reshape2)

 ggplot(data = Join_Maize, mapping = (aes(x=Chromosome, y=SNP_ID))) + geom_point()
 


 ggplot(data = Join_Teosinte, mapping = (aes(x=Chromosome, y=SNP_ID))) + geom_point()


# For distribution
 
 ggplot(data = Join_Maize) + geom_density(mapping = aes(x=SNP_ID), fill="blue")
 
 ggplot(data = Join_Teosinte) + geom_density(mapping = aes(x=SNP_ID), fill="blue")
           
 
# My Visulaization-Scatter plots
 
 plot(x=Join_Maize$Chromosome) 
 
 plot(x=Join_Teosinte$Chromosome)
 
 #Exported as Maize plot by using the export option in Plots pane. path is C:\Users\akhil\Desktop\R\R Assignment\Files\